src/Microsoft.OData.Core/ODataServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,72 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ODataServiceCollectionExtensions.cs" company="Microsoft">
+// Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    using System;
+    using Microsoft.OData;
+    using Microsoft.OData.Edm;
+    using Microsoft.OData.Json;
+    using Microsoft.OData.UriParser;
+
+    /// <summary>
+    /// Extension methods for <see cref="IServiceCollection"/>.
+    /// </summary>
+    public static class ODataServiceCollectionExtensions
+    {
+
+        /// <summary>
+        /// Adds the default OData required services to a given <see cref="IServiceCollection" />.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection" /> instance to add OData's default services to.</param>
+        /// <param name="odataVersion">An <see cref="ODataVersion" /> specifying which OData version to support. Defaults to <see cref="ODataVersion.V4" />.</param>
+        /// <param name="configureReaderAction">An <see cref="Action{ODataMessageReaderSettings}" /> that allows for configuring the default <see cref="ODataMessageReaderSettings" />.</param>
+        /// <param name="configureWriterAction">An <see cref="Action{ODataMessageWriterSettings}" /> that allows for configuring the default <see cref="ODataMessageWriterSettings" />.</param>
+        /// <param name="configureUriParserAction">An <see cref="Action{ODataUriParserSettings}" /> that allows for configuring the default <see cref="ODataUriParserSettings" />.</param>
+        /// <returns>The <see cref="IServiceCollection"/> instance we added OData services to, for fluent manipulation.</returns>
+        /// <remarks>This is usually called by OData platform libraries as a part of initializing runtime components of the OData platform.</remarks>
+        /// <exception cref="ArgumentNullException">Thrown if the <paramref name="services"/> parameter is <c>null</c>.</exception>
+        public static IServiceCollection AddDefaultODataServices(this IServiceCollection services,
+            ODataVersion odataVersion = ODataVersion.V4, 
+            Action<ODataMessageReaderSettings> configureReaderAction = default, 
+            Action<ODataMessageWriterSettings> configureWriterAction = default,
+            Action<ODataUriParserSettings> configureUriParserAction = default)

```

Comment:
I wonder whether it would be better to have one Options parameter that you can use to configure the different settings instead of having all these parameters. But maybe the options parameter would lead to confusion by introducing an extra layer of settings on top of a layer of settings, etc.

Reply:
Just to clarify, this is more of thinking-out-loud, not necessarily requesting a change.

Reply:
This is something that I've been trying to work out in my OData.NEXT thought experiment. Ultimately I think there should be a single Options object with multiple layers of settings that can be manipulated. That presently creates an N+1 problem that doesn't actually solve anything at the moment, so I went with this and optional parameters to give flexibility without creating something that will undoubtedly require a deeper plumbing change.